File: Code tham khảo/SIL/SilAdapterBasicCpp/CANoe/ApplicationModels/EnvironmentSimulation.cs
using System;
using Vector.Tools;
using Vector.CANoe.Runtime;
using Vector.CANoe.Sockets;
using Vector.CANoe.Threading;
using Vector.Diagnostics;
using NetworkDB;
using NetworkDB._SystemDataTypes;

using IServiceProvider = Vector.CANoe.Runtime.IServiceProvider;


public class EnvironmentSimulation : MeasurementScript
{
    // Counter for how many (timer) steps the average temperature is increased / decreased
    private static readonly int INCREASE_DECREASE_TEMP_COUNTER = 20;

    private int _increaseDecreaseTemperatureCounter;
    private bool _decreaseTemperature;
    private Timer _timer;
    private Random _rng;

    /// <summary>
    /// Called before measurement start to perform necessary initializations,
    /// e.g. to create objects. During measurement, few additional objects
    /// should be created to prevent garbage collection runs in time-critical
    /// simulations.
    /// </summary>
    public override void Initialize()
    {
      _increaseDecreaseTemperatureCounter = INCREASE_DECREASE_TEMP_COUNTER;
      _decreaseTemperature = true;
      _timer = new Timer(TimeSpan.FromMilliseconds(100), (sender, args) => IncreaseDecreaseTemperature());
      _rng = new Random();
    }
    
    /// <summary>Notification that the measurement starts.</summary>
    public override void Start()
    {
    
    }
    
    /// <summary>Notification that the measurement ends.</summary>
    public override void Stop()
    {
    
    }
    
    /// <summary>
    /// Cleanup after the measurement. Complement to Initialize. This is not
    /// a "Dispose" method; your object should still be usable afterwards.
    /// </summary>
    public override void Shutdown()
    {
    
    }

    [OnChange(Panel.Panel.MemberIDs.EnvironmentSimulationActive)]
    public void OnAutomaticModeChange()
    {
      if (Panel.Panel.EnvironmentSimulationActive.Value)
      {
        RoomTemperatureControl.Sensor1.Temperature.Value = 23.0;
        RoomTemperatureControl.Sensor2.Temperature.Value = 23.0;
        RoomTemperatureControl.Sensor3.Temperature.Value = 23.0;
        _timer.Start();
      }
      else
      {
        _timer.Stop();
      }
    }

    private void IncreaseDecreaseTemperature()
    {
        
      // If counter is still > 0, stay with increasing / decreasing temperature
      if (_increaseDecreaseTemperatureCounter > 0)
      {
        _increaseDecreaseTemperatureCounter -= 1;
      }
      else
      {
        // Counter for increasing / decreasing time interval / steps is 0, so
        // change from increasing to decreasing temperature or the other way around
        _decreaseTemperature = !_decreaseTemperature;
        _increaseDecreaseTemperatureCounter = INCREASE_DECREASE_TEMP_COUNTER;
      }

      var temperatureChange = _decreaseTemperature ? -1 : 1;
      var sensorNumber = _rng.Next(minValue: 1, maxValue: 4);
      switch (sensorNumber)
      {
        case 1:
          RoomTemperatureControl.Sensor1.Temperature.Value += temperatureChange;
          break;
        case 2:
          RoomTemperatureControl.Sensor2.Temperature.Value += temperatureChange;
          break;
        default:
          RoomTemperatureControl.Sensor3.Temperature.Value += temperatureChange;
          break;
      }
    }
}

File: Code tham khảo/SIL/SilAdapterBasicPython/CANoe/ApplicationModels/EnvironmentSimulation.py
import vector.canoe
from application_layer import RoomTemperatureControl
from application_layer.Panel import Panel
from random import choice


@vector.canoe.measurement_script
class EnvironmentSimulation:
    
    # Counter for how many (timer) steps the average temperature is increased / decreased
    INCREASE_DECREASE_TEMP_COUNTER = 20


    def __init__(self):
        self.increase_decrease_temperature_counter = self.INCREASE_DECREASE_TEMP_COUNTER
        self.decrease_temperature = True
        self.timer = vector.canoe.Timer(0.1, self.increase_decrease_temperature)

    @vector.canoe.on_change(Panel.EnvironmentSimulationActive)
    def on_automatic_mode_changed(self):
        if Panel.EnvironmentSimulationActive.copy() == True:
            RoomTemperatureControl.Sensor1.Temperature = 23
            RoomTemperatureControl.Sensor2.Temperature = 23
            RoomTemperatureControl.Sensor3.Temperature = 23
            self.timer.start()
        else:
            self.timer.cancel()


    def increase_decrease_temperature(self):
        # If counter is still > 0, stay with increasing/decreasing temperature
        if self.increase_decrease_temperature_counter > 0:
            self.increase_decrease_temperature_counter -= 1
        else:
            # Counter for increasing/decreasing time interval/steps is 0, so 
            # change from increasing to decreasing temperature or the other way round
            self.decrease_temperature = not self.decrease_temperature
            self.increase_decrease_temperature_counter = self.INCREASE_DECREASE_TEMP_COUNTER

        tempToChange = choice([RoomTemperatureControl.Sensor1, RoomTemperatureControl.Sensor2, RoomTemperatureControl.Sensor3])

        if self.decrease_temperature:
            tempToChange.Temperature = tempToChange.Temperature.copy() - 1
        else:
            tempToChange.Temperature = tempToChange.Temperature.copy() + 1

        self.timer = vector.canoe.Timer(0.1, self.increase_decrease_temperature)
        self.timer.start()
